File: emergency_tool.py
# Load
import time
import os
import platform
import urllib
import zipfile
import urllib.request

from route.tool.func import *

while True:
    data_db_load = input('Load DB (Y) [Y, N] : ')
    data_db_load = data_db_load.upper()
    if data_db_load in ('Y', 'N'):
        break

if data_db_load == 'Y':
    data_db_set = class_check_json()
    do_db_set(data_db_set)

    load_db = get_db_connect()

    conn = load_db.__enter__()
    curs = conn.cursor()
else:
    print('You can use [9, 11, 19]')

# Main
print('1. Backlink reset')
print('2. reCAPTCHA delete')
print('3. Ban delete')
print('4. Change host')
print('5. Change port')
print('6. Change skin')
print('7. Change password')
print('8. Set db version')
print('9. Delete set.json')
print('10. Change name')
print('11. Delete mysql.json')
print('14. Delete Main <HEAD>')
print('15. Give owner')
print('16. Delete 2FA password')
print('17. Change markup')
print('18. Change wiki access password')
print('19. Forced update')
print('20. Change domain')
print('21. Change TLS')
print('22. Delete body top')
print('23. Delete body bottom')
print('24. SQLite to MySQL')
print('25. Recalc exist data_set')
print('26. Change update branch')
print('27. Change golang port')

what_i_do = input('Insert selection number (EX : 9) : ')
if what_i_do == '1':
    try:
        go_num = int(input('Count (100) : '))
    except ValueError:
        go_num = 100

    curs.execute(db_change('select count(*) from data'))
    count = curs.fetchall()
    count = count[0][0] if count else 0

    num = 0
    title = []
    print('Load...')

    while num == 0 or len(title) == 100:
        curs.execute(db_change('select title from data limit ?, 100'), [num])
        title = curs.fetchall()

        print('Rest : ' + str(count - num))
        print('Start : ' + title[0][0])
        time.sleep(1)

        for name in title:
            num += 1
            if num % go_num == 0:
                print(str(num) + ' : ' + name[0])

            curs.execute(db_change("select data from data where title = ?"), [name[0]])
            data = curs.fetchall()

            render_lang_data = {
                'toc' : 'toc',
                'category' : 'category'
            }

            class_do_render(conn, render_lang_data, '').do_render(name[0], data[0][0], 'backlink')
elif what_i_do == '2':
    curs.execute(db_change("delete from other where name = 'recaptcha'"))
    curs.execute(db_change("delete from other where name = 'sec_re'"))
elif what_i_do == '3':
    user_data = input('IP or Name : ')

    curs.execute(
        db_change(
            "insert into rb (block, end, today, blocker, why, band, ongoing, login) "
            "values (?, ?, ?, ?, ?, ?, '', '')"
        ),
        [
            user_data,
            'release',
            get_time(),
            'tool:emergency',
            '',
            '',
        ]
    )

    curs.execute(db_change("update rb set ongoing = '' where block = ?"), [user_data])
elif what_i_do == '4':
    host = input('Host : ')

    curs.execute(db_change("update other set data = ? where name = 'host'"), [host])
elif what_i_do == '5':
    port = int(input('Port : '))

    curs.execute(db_change("update other set data = ? where name = 'port'"), [port])
elif what_i_do == '6':
    skin = input('Skin name : ')

    curs.execute(db_change("update other set data = ? where name = 'skin'"), [skin])
elif what_i_do == '7':
    user_name = input('User name : ')
    user_pw = input('User password : ')

    hashed = pw_encode(conn, user_pw)
    curs.execute(db_change("update user_set set data = ? where id = ? and name = 'pw'"), [
        hashed,
        user_name
    ])
elif what_i_do == '8':
    new_ver = input('Insert version (0000000) : ')

    if new_ver == '':
        new_ver = '0000000'

    curs.execute(db_change("update other set data = ? where name = 'ver'"), [new_ver])
elif what_i_do == '9':
    if os.path.exists(os.path.join('data', 'set.json')):
        os.remove(os.path.join('data', 'set.json'))
elif what_i_do == '10':
    user_name = input('User name : ')
    new_name = input('New name : ')

    curs.execute(
        db_change("update user_set set id = ? where id = ?"),
        [new_name, user_name]
    )
elif what_i_do == '11':
    if os.path.exists(os.path.join('data', 'mysql.json')):
        os.remove(os.path.join('data', 'mysql.json'))
elif what_i_do == '14':
    curs.execute(db_change('delete from other where name = "head"'))
elif what_i_do == '15':
    user_name = input('User name : ')

    curs.execute(db_change("update user_set set data = 'owner' where id = ? and name = 'acl'"), [user_name])
elif what_i_do == '16':
    user_name = input('User name : ')

    curs.execute(db_change('select data from user_set where name = "2fa" and id = ?'), [user_name])
    if curs.fetchall():
        curs.execute(db_change("update user_set set data = '' where name = '2fa' and id = ?"), [user_name])
elif what_i_do == '17':
    markup = input('Markup name : ')

    curs.execute(db_change("update other set data = ? where name = 'markup'"), [markup])
elif what_i_do == '18':
    wiki_access_password = input('Password : ')

    curs.execute(db_change("update other set data = ? where name = 'wiki_access_password'"), [wiki_access_password])
elif what_i_do == '19':
    curs.execute(db_change('select data from other where name = "update"'))
    up_data = curs.fetchall()
    up_data = up_data[0][0] if up_data and up_data[0][0] in ['stable', 'beta', 'dev', 'dont_use'] else 'stable'

    if platform.system() == 'Linux' or platform.system() == 'Darwin':
        ok = []
        ok += [os.system('git remote rm origin')]
        ok += [os.system('git remote add origin https://github.com/opennamu/opennamu.git')]
        ok += [os.system('git fetch --depth=1 origin ' + up_data)]
        ok += [os.system('git reset --hard origin/' + up_data)]
        for for_a in ok[1:]:
            if for_a != 0:
                print('Error : update failed')
                
                break
    elif platform.system() == 'Windows':
        os.system('rd /s /q route')
        
        urllib.request.urlretrieve('https://github.com/opennamu/opennamu/archive/' + up_data + '.zip', 'update.zip')
        
        zipfile.ZipFile('update.zip').extractall('')
        
        ok = os.system('xcopy /y /s /r opennamu-' + up_data + ' .')
        if ok == 0:
            os.system('rd /s /q opennamu-' + up_data)
            os.system('del update.zip')
        else:
            print('Error : update failed')
elif what_i_do == '20':
    domain = input('Domain (EX : 2du.pythonanywhere.com) : ')

    curs.execute(db_change('delete from other where name = "domain"'))
    curs.execute(db_change('insert into other (name, data, coverage) values ("domain", ?, "")'), [domain])
elif what_i_do == '21':
    tls_v = input('TLS (http) [http, https] : ')
    if not tls_v in ['http', 'https']:
        tls_v = 'http'

    curs.execute(db_change('delete from other where name = "http_select"'))
    curs.execute(db_change('insert into other (name, data, coverage) values ("http_select", ?, "")'), [tls_v])
elif what_i_do == '22':
    curs.execute(db_change('delete from other where name = "body"'))
elif what_i_do == '23':
    curs.execute(db_change('delete from other where name = "bottom_body"'))
elif what_i_do == '24':
    load_db = get_db_connect('mysql')
    mysql_conn = load_db.__enter__()
    mysql_curs = mysql_conn.cursor()
    
    load_db = get_db_connect('sqlite')
    sqlite_conn = load_db.__enter__()
    sqlite_curs = sqlite_conn.cursor()

    create_data = get_db_table_list()
    for create_table in create_data:
        create = ['test'] + create_data[create_table]

        create_r = ', '.join(['%s' for _ in create])
        create = ', '.join(create)
        
        mysql_curs.execute(db_change('delete from ' + create_table))
        
        sqlite_curs.execute(db_change('select ' + create + ' from ' + create_table))
        db_data = sqlite_curs.fetchall()
        if db_data:
            mysql_curs.executemany("insert into " + create_table + " (" + create + ") values (" + create_r + ")", db_data)
elif what_i_do == '25':
    curs.execute(db_change("select distinct doc_name from data_set where doc_rev = 'not_exist' or doc_rev = ''"))
    for for_a in curs.fetchall():
        data_set_exist = ''
        
        curs.execute(db_change("select title from data where title = ?"), [for_a[0]])
        if not curs.fetchall():
            data_set_exist = 'not_exist'

        curs.execute(db_change("update data_set set doc_rev = ? where doc_name = ? and (doc_rev = '' or doc_rev = 'not_exist')"), [data_set_exist, for_a[0]])
elif what_i_do == '26':
    up_data = input('Insert branch name (beta) [stable, beta, dev] : ')

    if not up_data in ['stable', 'beta', 'dev', 'dont_use']:
        up_data = 'beta'

    curs.execute(db_change('delete from other where name = "update"'))
    curs.execute(db_change('insert into other (name, data, coverage) values ("update", ?, "")'), [up_data])
elif what_i_do == '27':
    port = int(input('Port : '))

    curs.execute(db_change("update other set data = ? where name = 'golang_port'"), [port])
else:
    raise ValueError(what_i_do)

if data_db_load == 'Y':
    try:
        conn.commit()
    except:
        pass

print('OK')

File: route/emergency_tool.go
package route

import (
	"bufio"
	"database/sql"
	"errors"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"runtime"
	"strconv"
	"strings"

	"opennamu/route/tool"
	"opennamu/route/tool/markup"
)

func emergency_print_menu() {
	fmt.Println("1. Backlink reset")
	fmt.Println("2. reCAPTCHA delete")
	fmt.Println("3. Ban delete")
	fmt.Println("4. Change host")
	fmt.Println("5. Change port")
	fmt.Println("6. Change skin")
	fmt.Println("7. Change password")
	fmt.Println("8. Set db version")
	fmt.Println("9. Delete set.json")
	fmt.Println("10. Change name")
	fmt.Println("11. Delete mysql.json")
	fmt.Println("14. Delete Main <HEAD>")
	fmt.Println("15. Give owner")
	fmt.Println("16. Delete 2FA password")
	fmt.Println("17. Change markup")
	fmt.Println("18. Change wiki access password")
	fmt.Println("19. Forced binary update")
	fmt.Println("20. Change domain")
	fmt.Println("21. Change TLS")
	fmt.Println("22. Delete body top")
	fmt.Println("23. Delete body bottom")
	fmt.Println("24. SQLite to MySQL (not supported)")
	fmt.Println("25. Recalc exist data_set")
	fmt.Println("26. Change update branch")
	fmt.Println("27. Change golang port")
}

func emergency_input(reader *bufio.Reader, message string) string {
	fmt.Print(message)
	data, _ := reader.ReadString('\n')
	return strings.TrimSpace(data)
}

func emergency_open_db() (db *sql.DB, err error) {
	defer func() {
		if value := recover(); value != nil {
			err = fmt.Errorf("database open failed: %v", value)
		}
	}()

	tool.DB_boot()
	db = tool.DB_connect()
	if err := db.Ping(); err != nil {
		db.Close()
		return nil, err
	}

	return db, nil
}

func emergency_exec(db *sql.DB, query string, values ...any) error {
	_, err := db.Exec(tool.DB_change(query), values...)
	return err
}

func emergency_set_other(db *sql.DB, name string, data string) error {
	if err := emergency_exec(db, "update other set data = ? where name = ?", data, name); err != nil {
		return err
	}

	var exists int
	err := db.QueryRow(
		tool.DB_change("select 1 from other where name = ? limit 1"),
		name,
	).Scan(&exists)
	if err == nil {
		return nil
	}
	if !errors.Is(err, sql.ErrNoRows) {
		return err
	}

	return emergency_exec(
		db,
		"insert into other (name, data, coverage) values (?, ?, '')",
		name,
		data,
	)
}

func emergency_delete_file(path string) error {
	err := os.Remove(path)
	if errors.Is(err, os.ErrNotExist) {
		return nil
	}
	return err
}

func emergency_reset_backlink(db *sql.DB) error {
	if err := emergency_exec(db, "delete from back"); err != nil {
		return err
	}
	if err := emergency_exec(db, "delete from data_set where set_name = 'link_count'"); err != nil {
		return err
	}

	rows, err := db.Query(
		tool.DB_change("select title, coalesce(data, '') from data order by title"),
	)
	if err != nil {
		return err
	}

	documents := make([][]string, 0)
	for rows.Next() {
		var title string
		var data string
		if err := rows.Scan(&title, &data); err != nil {
			rows.Close()
			return err
		}
		documents = append(documents, []string{title, data})
	}
	if err := rows.Err(); err != nil {
		rows.Close()
		return err
	}
	rows.Close()

	error_count := 0
	for index, document := range documents {
		if err := emergency_render_backlink(db, document[0], document[1]); err != nil {
			error_count++
			fmt.Println("backlink error:", document[0], err)
		}

		if (index+1)%100 == 0 {
			fmt.Printf("Backlink reset: %d/%d\n", index+1, len(documents))
		}
	}

	fmt.Printf("Backlink reset: %d documents", len(documents))
	if error_count > 0 {
		fmt.Printf(" (%d errors)", error_count)
	}
	fmt.Println()
	return nil
}

func emergency_render_backlink(db *sql.DB, title string, data string) (err error) {
	defer func() {
		if value := recover(); value != nil {
			err = fmt.Errorf("%v", value)
		}
	}()

	markup.Get_render(db, title, data, "backlink")
	return nil
}

func emergency_recalc_data_set(db *sql.DB) error {
	rows, err := db.Query(
		tool.DB_change("select distinct doc_name from data_set where doc_rev = 'not_exist' or doc_rev = ''"),
	)
	if err != nil {
		return err
	}

	document_names := make([]string, 0)
	for rows.Next() {
		var document_name string
		if err := rows.Scan(&document_name); err != nil {
			rows.Close()
			return err
		}
		document_names = append(document_names, document_name)
	}
	if err := rows.Err(); err != nil {
		rows.Close()
		return err
	}
	rows.Close()

	for _, document_name := range document_names {
		var exists string
		err := db.QueryRow(
			tool.DB_change("select title from data where title = ? limit 1"),
			document_name,
		).Scan(&exists)
		if errors.Is(err, sql.ErrNoRows) {
			exists = "not_exist"
		} else if err != nil {
			return err
		} else {
			exists = ""
		}

		if err := emergency_exec(
			db,
			"update data_set set doc_rev = ? where doc_name = ? and (doc_rev = '' or doc_rev = 'not_exist')",
			exists,
			document_name,
		); err != nil {
			return err
		}
	}

	fmt.Printf("data_set recalculated: %d documents\n", len(document_names))
	return nil
}

func emergency_update_binary(branch string) error {
	if branch != "stable" && branch != "beta" {
		return fmt.Errorf("unsupported update branch: %s", branch)
	}

	executable, err := current_server_executable()
	if err != nil {
		return err
	}

	binary_url, err := get_server_update_url(branch)
	if err != nil {
		return err
	}

	temporary, err := os.CreateTemp(filepath.Dir(executable), ".opennamu-emergency-update-*")
	if err != nil {
		return err
	}
	temporary_path := temporary.Name()
	if err := temporary.Close(); err != nil {
		os.Remove(temporary_path)
		return err
	}

	if err := download_server_update(binary_url, temporary_path); err != nil {
		os.Remove(temporary_path)
		return err
	}

	if runtime.GOOS == "windows" {
		command := exec.Command(executable, server_update_mode, executable, temporary_path)
		detach_server_process(command)
		command.Stdout = os.Stdout
		command.Stderr = os.Stderr
		if err := command.Start(); err != nil {
			os.Remove(temporary_path)
			return err
		}
		fmt.Println("Update is scheduled. Close this process to replace the binary.")
		return nil
	}

	if err := os.Rename(temporary_path, executable); err != nil {
		os.Remove(temporary_path)
		return err
	}

	fmt.Println("Binary updated. Start openNAMU again.")
	return nil
}

func Run_emergency_tool(arguments []string) int {
	reader := bufio.NewReader(os.Stdin)
	choice := ""
	if len(arguments) > 0 {
		choice = arguments[0]
	} else {
		emergency_print_menu()
		choice = emergency_input(reader, "Insert selection number (EX : 9) : ")
	}

	if choice == "9" {
		if err := emergency_delete_file(filepath.Join("data", "set.json")); err != nil {
			fmt.Fprintln(os.Stderr, "Emergency tool failed:", err)
			return 1
		}
		fmt.Println("OK")
		return 0
	}
	if choice == "11" {
		if err := emergency_delete_file(filepath.Join("data", "mysql.json")); err != nil {
			fmt.Fprintln(os.Stderr, "Emergency tool failed:", err)
			return 1
		}
		fmt.Println("OK")
		return 0
	}
	if choice == "19" {
		branch := emergency_input(reader, "Branch (stable/beta) : ")
		if branch == "" {
			branch = "stable"
		}
		if err := emergency_update_binary(branch); err != nil {
			fmt.Fprintln(os.Stderr, "Emergency tool failed:", err)
			return 1
		}
		return 0
	}
	if choice == "24" {
		fmt.Fprintln(os.Stderr, "SQLite to MySQL conversion is not supported in the Go emergency tool.")
		return 1
	}

	db, err := emergency_open_db()
	if err != nil {
		fmt.Fprintln(os.Stderr, "Emergency tool failed:", err)
		return 1
	}
	defer tool.DB_close(db)

	switch choice {
	case "1":
		err = emergency_reset_backlink(db)
	case "2":
		err = emergency_exec(db, "delete from other where name = 'recaptcha'")
		if err == nil {
			err = emergency_exec(db, "delete from other where name = 'sec_re'")
		}
	case "3":
		user_data := emergency_input(reader, "IP or Name : ")
		err = emergency_exec(
			db,
			"insert into rb (block, end, today, blocker, why, band, ongoing, login) values (?, 'release', ?, 'tool:emergency', '', '', '', '')",
			user_data,
			tool.Get_time(),
		)
		if err == nil {
			err = emergency_exec(db, "update rb set ongoing = '' where block = ?", user_data)
		}
	case "4":
		err = emergency_set_other(db, "host", emergency_input(reader, "Host : "))
	case "5":
		port := emergency_input(reader, "Port : ")
		port_num, port_err := strconv.Atoi(port)
		if port_err != nil || port_num < 1 || port_num > 65535 {
			err = fmt.Errorf("invalid port: %s", port)
		} else {
			err = emergency_set_other(db, "port", port)
		}
	case "6":
		err = emergency_set_other(db, "skin", emergency_input(reader, "Skin name : "))
	case "7":
		user_name := emergency_input(reader, "User name : ")
		user_password := emergency_input(reader, "User password : ")
		password_encode := tool.Get_user_encode(db, user_name)
		password := tool.Password_encode(db, user_password, password_encode)
		err = emergency_exec(
			db,
			"update user_set set data = ? where id = ? and name = 'pw'",
			password,
			user_name,
		)
	case "8":
		version := emergency_input(reader, "Insert version (0000000) : ")
		if version == "" {
			version = "0000000"
		}
		err = emergency_set_other(db, "ver", version)
	case "10":
		user_name := emergency_input(reader, "User name : ")
		new_name := emergency_input(reader, "New name : ")
		err = emergency_exec(db, "update user_set set id = ? where id = ?", new_name, user_name)
	case "14":
		err = emergency_exec(db, "delete from other where name = 'head'")
	case "15":
		user_name := emergency_input(reader, "User name : ")
		var user_exists string
		err = db.QueryRow(
			tool.DB_change("select id from user_set where id = ? limit 1"),
			user_name,
		).Scan(&user_exists)
		if errors.Is(err, sql.ErrNoRows) {
			err = fmt.Errorf("user not found: %s", user_name)
		}
		if err == nil {
			err = emergency_exec(db, "update user_set set data = 'owner' where id = ? and name = 'acl'", user_name)
		}
		if err == nil {
			var acl_exists int
			acl_err := db.QueryRow(
				tool.DB_change("select 1 from user_set where id = ? and name = 'acl' limit 1"),
				user_name,
			).Scan(&acl_exists)
			if errors.Is(acl_err, sql.ErrNoRows) {
				err = emergency_exec(db, "insert into user_set (name, id, data) values ('acl', ?, 'owner')", user_name)
			} else {
				err = acl_err
			}
		}
	case "16":
		user_name := emergency_input(reader, "User name : ")
		err = emergency_exec(db, "update user_set set data = '' where name = '2fa' and id = ?", user_name)
	case "17":
		err = emergency_set_other(db, "markup", emergency_input(reader, "Markup name : "))
	case "18":
		err = emergency_set_other(db, "wiki_access_password", emergency_input(reader, "Password : "))
	case "20":
		err = emergency_set_other(db, "domain", emergency_input(reader, "Domain (EX : 2du.pythonanywhere.com) : "))
	case "21":
		tls_value := emergency_input(reader, "TLS (http) [http, https] : ")
		if tls_value != "https" {
			tls_value = "http"
		}
		err = emergency_set_other(db, "http_select", tls_value)
	case "22":
		err = emergency_exec(db, "delete from other where name = 'body'")
	case "23":
		err = emergency_exec(db, "delete from other where name = 'bottom_body'")
	case "25":
		err = emergency_recalc_data_set(db)
	case "26":
		branch := emergency_input(reader, "Insert branch name (beta) [stable, beta] : ")
		if branch != "stable" && branch != "beta" {
			branch = "beta"
		}
		err = emergency_set_other(db, "update", branch)
	case "27":
		port := emergency_input(reader, "Port : ")
		port_num, port_err := strconv.Atoi(port)
		if port_err != nil || port_num < 1 || port_num > 65535 {
			err = fmt.Errorf("invalid port: %s", port)
		} else {
			err = emergency_set_other(db, "golang_port", port)
		}
	default:
		err = fmt.Errorf("unknown selection: %s", choice)
	}

	if err != nil {
		fmt.Fprintln(os.Stderr, "Emergency tool failed:", err)
		return 1
	}

	fmt.Println("OK")
	return 0
}
